### ISSUE TITLE
Bug url cut on line returns

### ISSUE THREAD

**Issue by lefablab** [PHPMailer/PHPMailer#2227]
Hello,

Following a change of web hosting and PHP version, I switched from version 5.2.8 to version 5.5 of PHPMailer.

A problem appeared: some urls in the HTML version of the sent messages are modified.
This seems to be related to line breaks.

Here is an extract of the source code of a sample email containing an erroneous url:
```
                   <td valign=3D"top">
                    <p style=3D"font-family:'Avenir Next','Helvetica Neue',=
 Helvetica, Arial, sans-serif;font-size: 16px;line-height: 1.5em;margin: 0 =
0 20px 0;font-weight: 400;color: #303030;">[Abondance.] Goossips : Regex et=
 Search Console, D=C3=A9sinformation, Paywall <a style=3D"text-decoration:n=
one;color:#007793;font-weight:bold;letter-spacing:1px;" href=3D'https://www=
..hellofacteur.com/go/4504/5388015/3435' title=3D"Publi=C3=A9 dans : Abondan=
ce.">+ LIRE</a>.</p>                   </td>
```

A "." is added after the line break which gives an erroneous url starting with 'https://www..hellofacteur.com' instead of 'https://www.hellofacteur.com'. This url is not present in the HTML code generated by my application.
Moreover, the DKIM Verifier extension of Thunderbird tells me that "The email is corrupted". Emails are sent via mailjet, which was already the case before I updated PHPMailer.

This is not systematic. Some mails are ok. I think this is caused by urls being cut in the wrong place due to line feeds. I don't understand why a dot is added after the line break?

Here is what the scripts called by my application to send mails look like:

`$mail=new PHPMailer(true);
$mail->isSMTP();
$mail->Host=SMTP_SERVEUR;
$mail->SMTPAuth=true;
$mail->Username=SMTP_NAME;
$mail->Password=SMTP_PASSE;
$mail->Port=SMTP_PORT;
$mail->SMTPSecure=SMTP_SECURE;
$mail->setLanguage(PHPMAIL_LANG);
$mail->CharSet="UTF-8";
$mail->Subject=$Mail_sujet;
$mail->isHTML(true);
$mail->Body=$Mail_boby_html;
$mail->AltBody=$Mail_boby_txt;
$mail->addReplyTo($Mail_reponse_email,$Mail_reponse_nom);
$mail->FromName=$Mail_expediteur_nom;
$mail->From=$Mail_expediteur_email;
$mail->addAddress($Mail_destinataire_email,$Mail_destinataire_nom);`

Is this a known problem with PHPMailer? Thank you.

**Comment by Synchro** [comment]
There is no such version as PHPMailer 5.5? Are you seeing this on PHPMailer 6.2.0 with PHP 7.4 or 8.0?

Adding dots after line breaks is most likely due to something called dot stuffing, which is a required part of [RFC5321](https://tools.ietf.org/html/rfc5321#section-4.5.2). In SMTP, a dot on a line by itself is used to denote the end of a message, so if a line within a message body contains that sequence, an additional dot is added; receivers know to remove the additional dot. This can occur at any point in a message where there happens to be a dot following a line break, it's not something specific to URLs, and it should be transparent to the sender.

**Comment by lefablab** [comment]
It's PHPMailer 6.2.0, sorry !

So the problem could be Thunderbird? But I don't see it with other emails. And with each error, the extra dot is at the beginning of the line, following a return to the line. This does not only concern the urls. In the following code, the final point is doubled in the display :

```
                   <td valign=3D"top">
                    <p style=3D"font-family:'Avenir Next','Helvetica Neue',=
 Helvetica, Arial, sans-serif;font-size: 16px;line-height: 1.5em;margin: 0 =
0 20px 0;font-weight: 400;color: #303030;">[Developpez.com] Microsoft pr=C3=
=A9voit de mettre fin au mot de passe en 2021 pour s'appuyer sur de nouvell=
es m=C3=A9thodes d'authentification comme Windows Hello, Microsoft Authenti=
cator et la biom=C3=A9trie <a style=3D"text-decoration:none;color:#007793;f=
ont-weight:bold;letter-spacing:1px;" href=3D'https://www.hellofacteur.com/g=
o/4504/5388026/3435' title=3D"Publi=C3=A9 dans : Developpez.com">+ LIRE</a>=
..</p>                   </td>
```

**Comment by lefablab** [comment]
I just tested the opening with the Gmail webmail and the problem is the same. The two points of the url are not corrected. So it's not specific to Thunderbird. Same thing with K-9 Mail on Android.

**Comment by Synchro** [comment]
Are you by any chance sending *from* a Microsoft Exchange server? In the past I have seen it cause problems relating to dot stuffing.

**Comment by lefablab** [comment]
I use maijet's servers to send the messages. This was already the case before my change of web hosting, so I don't think it comes from that.

I read in the PHPMailer code that it had message corruption problems with some versions of PHP7 (function preSend()). This morning I changed the php version from the default (7.3.18) to 7.4.12. I'll see if that changes anything.

The problem is difficult to reproduce voluntarily because it occurs "randomly", when the added dot is at the beginning of a new line. So some mails are ok...

**Comment by Synchro** [comment]
What it's doing is correct behaviour, [the implementation is as I would expect it to be](https://github.com/PHPMailer/PHPMailer/blob/master/src/SMTP.php#L769). If receivers are handling it incorrectly, that's not the sender's fault.

If receivers are not handling this properly, you should be able to provoke unexpected behaviour in them by having some lines in your message body that are just a `.`, and I would expect it to truncate the message at that point. Without dot-stuffing, the dot is indistinguishable from an "end of DATA" marker. It might be interesting to try that.

**Comment by lefablab** [comment]
I did some tests. When the messages are sent directly to Gmail, no problem appears.
So either the problem comes from the e-mail client or gmail fixes the problem.

But from what I understand from section 4.5.2 of RFC2821, it's the SMTP server that has to remove the duplicate dots.
So here the error would come from mailjet, Thunderbird just interpreting the received code without modifying it.

Moreover, I tried to manually add a link and the source code of the received message is ok, the second point being deleted when Thunderbird receives it :

`                  <td valign=3D"top">
                    <p style=3D"font-family:'Avenir Next','Helvetica Neue',=
 Helvetica, Arial, sans-serif;font-size: 16px;line-height: 1.5em;margin: 0 =
0 20px 0;font-weight: 400;color: #303030;">................................=
.......................................<a style=3D"text-decoration:none;col=
or:#007793;font-weight:bold;letter-spacing:1px;" href=3D'http://www.example=
.com'>clic</a>                   </td>`

This is not the case for messages containing erroneous links :
`                   <td valign=3D"top">
                    <p style=3D"font-family:'Avenir Next','Helvetica Neue',=
 Helvetica, Arial, sans-serif;font-size: 16px;line-height: 1.5em;margin: 0 =
0 20px 0;font-weight: 400;color: #303030;">[Ouest France] A Kerpape, le Med=
ef joue le P=C3=A8re No=C3=ABl - Ouest-France <a style=3D"text-decoration:n=
one;color:#007793;font-weight:bold;letter-spacing:1px;" href=3D'https://www=
..hellofacteur.com/go/3937/5390528/1008' title=3D"Publi=C3=A9 dans : Ouest-F=
rance : Morbihan (56).">+ LIRE</a>.</p>                   </td>`

There must be something in the messages sent by my application that disrupts the operation of mailjet in some cases. I use a free account and cannot ask for their support.
